### PR TITLE
Fix stack depth and memory issues related to types

### LINKED ISSUE #1276: Guard large allocation in Value::deserialize, apply recursive limits on types
Malformed byte strings in Clarity deserialization can cause OOM.

### PR DESCRIPTION
**Pull request by kantai**
This PR resolves #1276, and other stack overflow conditions coming from general AST and type depths.

1. Adds a maximum type depth to all types
2. Explicit enforcement on `Value::deserialize_read`
3. Use bound reader for value deserialization
4. Enforce size and depth limits on some/err/okay type construction
5. Add an AST pass for checking AST depth

Tests for the AST depth checker, the size and depth enforcements, and the deserialization enforcement.

### CODE REVIEW

File: src/util/retry.rs [hunk 1]
```diff
@@ -103,11 +103,10 @@ impl<'a, R: Read> BoundReader<'a, R> {
 
 impl <'a, R: Read> Read for BoundReader<'a, R> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        if self.read_so_far.checked_add(buf.len() as u64).is_none() {
-            return Err(io::Error::new(io::ErrorKind::Other, "Read would overflow u64".to_string()));
-        }
+        let intended_read = self.read_so_far.checked_add(buf.len() as u64)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Read would overflow u64".to_string()))?;
```
Comment (jcnelson) [inline src/util/retry.rs]: God help us if this ever gets reached, but thanks for adding this :+1:

File: src/vm/mod.rs [hunk 2]
```diff
@@ -284,6 +285,26 @@ mod test {
...
+                        (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ 
+                       bar 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1)
+                         1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1))
+                       (define-private (bar) 
+                        (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ 
+                        (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ 
+                       1 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1)
+                         1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1))
+                       (foo)
+                      ";
```
Comment (lgalabru) [inline src/vm/mod.rs]: Can we also include a test that would do a `contract-call!` (static and dynamic), instead of internal call ?

Comment (kantai) [reply src/vm/mod.rs]: Sure -- added in cfa22148e

Comment (lgalabru) [reply src/vm/mod.rs]: Oh I think that https://github.com/blockstack/blockstack-core/pull/1301/commits/cfa22148e631216b47af6004af19543a7b7ccca4 was pushed on the other branch, but this is probably fine :)

Comment (kantai) [reply src/vm/mod.rs]: Yep, oops :(

I'm merging that into this branch anyways, so it's fine.

**Pull request merged**